### ISSUE TITLE
Crux failed whilst running performance tests.

### ISSUE THREAD

**Issue by acquayefrank** [crux-toolkit/crux-toolkit#528]
Hello, @CharlesEGrant.
After building crux on Ubuntu, I run the performance test and it failed due to a static normalizer error.
I have added the stack trace below.
<pre>
RUN: ../../src/crux tide-index --output-dir worm+contaminants --parameter-file crux.param worm+contaminants.fa worm+contaminants
INFO: Writing results to output directory 'worm+contaminants'.
INFO: CPU: L429746
INFO: Crux version: 4.1-03be191d-2021-10-13
INFO: Пт окт 22 13:45:01 MSK 2021
INFO: Beginning tide-index.
INFO: Running tide-index...
WARNING: The output directory 'worm+contaminants' already exists.
Existing files will be overwritten.
INFO: Reading worm+contaminants.fa and computing unmodified peptides...
INFO: Generated 584624 targets, including duplicates.
INFO: Failed to generate decoys for 232 low complexity peptides.
INFO: Generated 452303 decoys.
INFO: Writing decoy fasta...
INFO: Generating 1 decoy per target
INFO: Reading proteins
INFO: Wrote 100000 peptides
INFO: Wrote 200000 peptides
INFO: Wrote 300000 peptides
INFO: Wrote 400000 peptides
INFO: Wrote 500000 peptides
INFO: Wrote 600000 peptides
INFO: Wrote 700000 peptides
INFO: Wrote 800000 peptides
INFO: Wrote 900000 peptides
INFO: Skipped 132089 duplicate targets and 143 duplicate decoys.
INFO: Wrote 452535 targets and 452160 decoys.
INFO: Writing peptide lists...
INFO: Precomputing theoretical spectra...
INFO: Elapsed time: 20.4 s
INFO: Finished crux tide-index.
INFO: Return Code:0
INFO: Ion cache check in: 0 check out:0
RUN: ../../src/crux tide-search --output-dir tide-xcorr --parameter-file crux.param  051708-worm-ASMS-10.ms2 worm+contaminants
INFO: Writing results to output directory 'tide-xcorr'.
INFO: CPU: L429746
INFO: Crux version: 4.1-03be191d-2021-10-13
INFO: Пт окт 22 13:45:21 MSK 2021
INFO: Beginning tide-search.
INFO: Running tide-search...
INFO: Number of Threads: 1
INFO: Reading index worm+contaminants
INFO: Read 23932 target proteins
INFO: Converting 051708-worm-ASMS-10.ms2 to spectrumrecords format
INFO: Elapsed time starting conversion: 0.52 s
INFO: Converting ms_level 2 ...
INFO: Reading spectrum file tide-xcorr/051708-worm-ASMS-10.ms2.spectrumrecords.tmp.
INFO: Read 14298 spectra.
INFO: Starting search.
INFO: 1000 spectrum-charge combinations searched, 7% complete
INFO: 2000 spectrum-charge combinations searched, 14% complete
INFO: 3000 spectrum-charge combinations searched, 21% complete
INFO: 4000 spectrum-charge combinations searched, 28% complete
INFO: 5000 spectrum-charge combinations searched, 35% complete
INFO: 6000 spectrum-charge combinations searched, 42% complete
INFO: 7000 spectrum-charge combinations searched, 49% complete
INFO: 8000 spectrum-charge combinations searched, 56% complete
INFO: 9000 spectrum-charge combinations searched, 63% complete
INFO: 10000 spectrum-charge combinations searched, 70% complete
INFO: 11000 spectrum-charge combinations searched, 77% complete
INFO: 12000 spectrum-charge combinations searched, 84% complete
INFO: 13000 spectrum-charge combinations searched, 91% complete
INFO: 14000 spectrum-charge combinations searched, 98% complete
INFO: [Thread 0]: Deleted 131636 precursor, 0 isotope and 126896 out-of-range peaks.
INFO: [Thread 0]: Retained 94.0339% of peaks.
INFO: Time per spectrum-charge combination: 0.006646 s.
INFO: Average number of candidates per spectrum-charge combination: 1922.031616
INFO: Elapsed time: 95.1 s
INFO: Finished crux tide-search.
INFO: Return Code:0
INFO: Ion cache check in: 0 check out:0
RUN: ../../src/crux assign-confidence --output-dir tide-xcorr  tide-xcorr/tide-search.txt
WARNING: The output directory 'tide-xcorr' already exists.
Existing files will not be overwritten.
INFO: CPU: L429746
INFO: Crux version: 4.1-03be191d-2021-10-13
INFO: Пт окт 22 13:46:57 MSK 2021
INFO: Beginning assign-confidence.
INFO: Estimation method = tdc.
INFO: Assigning index 0 to 051708-worm-ASMS-10.ms2.
INFO: Found 71490 PSMs in tide-xcorr/tide-search.txt.
INFO: Automatically detected score type: xcorr_score
INFO: Score type=xcorr_score, sorting in descending order
INFO: Skipped 29284 target and 27908 decoy PSMs with rank > 1.
INFO: There are 8710 target and 5588 decoy PSMs for q-value computation.
INFO: Number of PSMs at 1% FDR = 1961.
INFO: Number of PSMs at 5% FDR = 2333.
INFO: Number of PSMs at 10% FDR = 2667.
INFO: Elapsed time: 2.57 s
INFO: Finished crux assign-confidence.
INFO: Return Code:0
INFO: Ion cache check in: 0 check out:0
RUN: ../../src/crux percolator --output-dir tide-xcorr tide-xcorr/tide-search.txt
WARNING: The output directory 'tide-xcorr' already exists.
Existing files will not be overwritten.
INFO: CPU: L429746
INFO: Crux version: 4.1-03be191d-2021-10-13
INFO: Пт окт 22 13:46:59 MSK 2021
INFO: Beginning percolator.
INFO: Converting input to pin format.
INFO: Parsing tide-xcorr/tide-search.txt
INFO: Assigning index 0 to 051708-worm-ASMS-10.ms2.
INFO: There are 37994 target matches and 33496 decoys
INFO: Maximum observed charge is 3.
INFO: File conversion complete.
INFO: Percolator version 3.05.nightly-137-e806a0c5, Build Date Oct 18 2021 14:11:08
INFO: Copyright (c) 2006-9 University of Washington. All rights reserved.
INFO: Written by Lukas Käll (lukall@u.washington.edu) in the
INFO: Department of Genome Sciences at the University of Washington.
INFO: Issued command:
INFO: percolator --results-peptides tide-xcorr/percolator.target.peptides.txt --decoy-results-peptides tide-xcorr/percolator.decoy.peptides.txt --results-psms tide-xcorr/percolator.target.psms.txt --decoy-results-psms tide-xcorr/percolator.decoy.psms.txt --verbose 2 --protein-decoy-pattern decoy_ --seed 1 --subset-max-train 0 --trainFDR 0.01 --testFDR 0.01 --maxiter 10 --search-input auto --no-schema-validation --protein-enzyme trypsin --post-processing-tdc tide-xcorr/make-pin.pin
INFO: Started Fri Oct 22 13:47:04 2021
INFO: Hyperparameters: selectionFdr=0.01, Cpos=0, Cneg=0, maxNiter=10
INFO: Reading tab-delimited input from datafile tide-xcorr/make-pin.pin
INFO: Features:
INFO: lnrSp deltLCn deltCn XCorr Sp IonFrac PepLen Charge1 Charge2 Charge3 enzN enzC enzInt lnNumDSP dM absdM
INFO: Found 71490 PSMs
INFO: Separate target and decoy search inputs detected, using target-decoy competition on Percolator scores.
INFO: Train/test set contains 37994 positives and 33496 negatives, size ratio=1.13428 and pi0=1
INFO: Selecting Cpos by cross-validation.
INFO: Selecting Cneg by cross-validation.
INFO: Split 1:  Selected feature 2 as initial direction. Could separate 1420 training set positives with q<0.01 in that direction.
INFO: Split 2:  Selected feature 2 as initial direction. Could separate 1452 training set positives with q<0.01 in that direction.
INFO: Split 3:  Selected feature 2 as initial direction. Could separate 1417 training set positives with q<0.01 in that direction.
INFO: Found 2126 test set positives with q<0.01 in initial direction
INFO: Reading in data and feature calculation took 4.4767 cpu seconds or 3 seconds wall clock time.
INFO: ---Training with Cpos selected by cross validation, Cneg selected by cross validation, initial_fdr=0.01, fdr=0.01
INFO: Iteration 1:      Estimated 2450 PSMs with q<0.01
INFO: Iteration 2:      Estimated 2482 PSMs with q<0.01
INFO: Iteration 3:      Estimated 2502 PSMs with q<0.01
INFO: Iteration 4:      Estimated 2516 PSMs with q<0.01
INFO: Iteration 5:      Estimated 2544 PSMs with q<0.01
INFO: Iteration 6:      Estimated 2547 PSMs with q<0.01
INFO: Iteration 7:      Estimated 2551 PSMs with q<0.01
INFO: Iteration 8:      Estimated 2551 PSMs with q<0.01
INFO: Iteration 9:      Estimated 2554 PSMs with q<0.01
INFO: Iteration 10:     Estimated 2555 PSMs with q<0.01
INFO: Learned normalized SVM weights for the 3 cross-validation splits:
INFO:  Split1    Split2  Split3 FeatureName
INFO: -1.0202   -1.3770 -0.8505 lnrSp
INFO:  0.4362    1.3939  1.0768 deltLCn
INFO:  0.3395    0.2432  0.2167 deltCn
INFO:  0.6920    0.2064  0.6620 XCorr
INFO:  0.0586    0.5635  0.3214 Sp
INFO:  0.1795    0.0731 -0.2363 IonFrac
INFO:  0.6711    1.9326  1.5799 PepLen
INFO:  0.1297   -0.1770  0.0581 Charge1
INFO:  0.2015    0.3375  0.1822 Charge2
INFO: -0.2288   -0.3003 -0.1944 Charge3
INFO:  0.2245    0.1939  0.2899 enzN
INFO:  0.0000    0.0000  0.0000 enzC
INFO: -0.2751   -0.3602 -0.5412 enzInt
INFO:  0.7305    2.1815  2.0107 lnNumDSP
INFO: -0.3358    0.7582  0.1855 dM
INFO: -0.6909   -1.1386 -0.8457 absdM
INFO: -5.6815   -7.4012 -6.0362 m0
INFO: Found 2388 test set PSMs with q<0.01.
INFO: Selected best-scoring PSM per scan+expMass (target-decoy competition): 8982 target PSMs and 5316 decoy PSMs.
crux: .../crux-toolkit/ext/build/src/percolator/src/Normalizer.cpp:71: static Normalizer* Normalizer::getNormalizer(): Assertion `false' failed.
Aborted
Child was terminated by signal -134
</pre>

**Comment by wsnoble** [comment]
If this is a change in the percolator code, we should post an issue to the percolator github. Can you go ahead and do that, to suggest this change?

**Comment by kfattila** [comment]
Can you reproduce this bug in your system by running the performance-tests with your latest version? I have a suspicion that it occurs only on our server. 
We agreed at our first meeting that Charles will take a look at this problem. It can be that make.pin program or something else in CRUX creates an instance of that Normalizer.

**Comment by wsnoble** [comment]
I can verify that this problem arises on our linux systems as well.

`crux: /net/noble/vol1/home/noble/proj/crux/crux-toolkit/build/ext/build/src/percolator/src/Normalizer.cpp:71: static Normalizer* Normalizer::getNormalizer(): Assertion `false' failed.`

Here is the internal percolator command that yielded the error:

`percolator --results-peptides tide-xcorr/percolator.target.peptides.txt --decoy-results-peptides tide-xcorr/percolator.decoy.peptides.txt --results-psms tide-xcorr/percolator.target.psms.txt --decoy-results-psms tide-xcorr/percolator.decoy.psms.txt --verbose 2 --protein-decoy-pattern decoy_ --seed 1 --subset-max-train 0 --trainFDR 0.01 --testFDR 0.01 --maxiter 10 --search-input auto --no-schema-validation --protein-enzyme trypsin --post-processing-tdc tide-xcorr/make-pin.pin`

I will try to reproduce this with the standalone percolator.

**Comment by wsnoble** [comment]
This worked fine with a random version of percolator I happened to have lying around.

3.02.nightly-3-ffdef88, Build Date Aug 27 2018

I tried to build the latest percolator, but that failed (https://github.com/percolator/percolator/issues/320)

**Comment by wsnoble** [comment]
I verified that this works fine with the latest version of Percolator, at least on a mac.  So whatever the problem is, it seems like it's in Crux.  :(

[make-pin.pin.gz](https://github.com/crux-toolkit/crux-toolkit/files/7663802/make-pin.pin.gz)
```
(base) wnoble@Williams-MacBook-Pro-2 Downloads % [?2004h~/software/percolator-github/build/src/percolator --results-peptides tide-xcorr/percolator.target.peptides.txt --decoy-results-peptides tide-xcorr/percolator.decoy.peptides.txt --results-psms tide-xcorr/percolator.target.psms.txt --decoy-results-psms tide-xcorr/percolator.decoy.psms.txt --verbose 2 --protein-decoy-pattern decoy_ --seed 1 --subset-max-train 0 --trainFDR 0.01 --testFDR 0.01 --maxiter 10 --search-input auto --no-schema-validation --protein-enzyme trypsin --post-processing-tdc make-pin.pin
<hema-validation --protein-enzyme trypsin --post-processing-tdc make-pin.pin[?2004l
Percolator version 3.05.nightly-137-e806a0c5, Build Date Dec  6 2021 10:02:35
Copyright (c) 2006-9 University of Washington. All rights reserved.
Written by Lukas Käll (lukall@u.washington.edu) in the
Department of Genome Sciences at the University of Washington.
Issued command:
/Users/wnoble/software/percolator-github/build/src/percolator --results-peptides tide-xcorr/percolator.target.peptides.txt --decoy-results-peptides tide-xcorr/percolator.decoy.peptides.txt --results-psms tide-xcorr/percolator.target.psms.txt --decoy-results-psms tide-xcorr/percolator.decoy.psms.txt --verbose 2 --protein-decoy-pattern decoy_ --seed 1 --subset-max-train 0 --trainFDR 0.01 --testFDR 0.01 --maxiter 10 --search-input auto --no-schema-validation --protein-enzyme trypsin --post-processing-tdc make-pin.pin
Started Mon Dec  6 12:27:03 2021
Hyperparameters: selectionFdr=0.01, Cpos=0, Cneg=0, maxNiter=10
Reading tab-delimited input from datafile make-pin.pin
Features:
lnrSp deltLCn deltCn XCorr Sp IonFrac PepLen Charge1 Charge2 Charge3 enzN enzC enzInt lnNumDSP dM absdM 
Found 71490 PSMs
Separate target and decoy search inputs detected, using target-decoy competition on Percolator scores.
Train/test set contains 37994 positives and 33496 negatives, size ratio=1.13428 and pi0=1
Selecting Cpos by cross-validation.
Selecting Cneg by cross-validation.
Split 1:	Selected feature 2 as initial direction. Could separate 1420 training set positives with q<0.01 in that direction.
Split 2:	Selected feature 2 as initial direction. Could separate 1452 training set positives with q<0.01 in that direction.
Split 3:	Selected feature 2 as initial direction. Could separate 1417 training set positives with q<0.01 in that direction.
Found 2126 test set positives with q<0.01 in initial direction
Reading in data and feature calculation took 0.7835 cpu seconds or 1 seconds wall clock time.
---Training with Cpos selected by cross validation, Cneg selected by cross validation, initial_fdr=0.01, fdr=0.01
Iteration 1:	Estimated 2450 PSMs with q<0.01
Iteration 2:	Estimated 2482 PSMs with q<0.01
Iteration 3:	Estimated 2502 PSMs with q<0.01
Iteration 4:	Estimated 2516 PSMs with q<0.01
Iteration 5:	Estimated 2544 PSMs with q<0.01
Iteration 6:	Estimated 2547 PSMs with q<0.01
Iteration 7:	Estimated 2551 PSMs with q<0.01
Iteration 8:	Estimated 2551 PSMs with q<0.01
Iteration 9:	Estimated 2554 PSMs with q<0.01
Iteration 10:	Estimated 2555 PSMs with q<0.01
Learned normalized SVM weights for the 3 cross-validation splits:
 Split1	 Split2	 Split3	FeatureName
-1.0202	-1.3770	-0.8505	lnrSp
 0.4362	 1.3939	 1.0768	deltLCn
 0.3395	 0.2432	 0.2167	deltCn
 0.6920	 0.2064	 0.6620	XCorr
 0.0586	 0.5635	 0.3214	Sp
 0.1795	 0.0731	-0.2363	IonFrac
 0.6711	 1.9326	 1.5799	PepLen
 0.1297	-0.1770	 0.0581	Charge1
 0.2015	 0.3375	 0.1822	Charge2
-0.2288	-0.3003	-0.1944	Charge3
 0.2245	 0.1939	 0.2899	enzN
 0.0000	 0.0000	 0.0000	enzC
-0.2751	-0.3602	-0.5412	enzInt
 0.7305	 2.1815	 2.0107	lnNumDSP
-0.3358	 0.7582	 0.1855	dM
-0.6909	-1.1386	-0.8457	absdM
-5.6815	-7.4012	-6.0362	m0
Found 2388 test set PSMs with q<0.01.
Selected best-scoring PSM per scan+expMass (target-decoy competition): 8982 target PSMs and 5316 decoy PSMs.
Tossing out "redundant" PSMs keeping only the best scoring PSM for each unique peptide.
Calculating q values.
Final list yields 1385 target peptides with q<0.01.
Calculating posterior error probabilities (PEPs).
Processing took 12.7784 cpu seconds or 13 seconds wall clock time.
```

**Comment by kfattila** [comment]
I can fix this easily.  The PercolatorAdapter.cpp creates the normalizer, all we need to do is just to free the normalizer and call the Percolator::resetNormalizer member function.

**Comment by kfattila** [comment]
I was wrong, it isn't easy. The normalizer is need to unnormalize the scores when converting Percolator's results into CRUX's match  collection. We have two ways to fix it:
1. Either we ask Lukas to remove the assert part, or
2.  Crux does not unnormalize Percolator's results.

**Comment by wsnoble** [comment]
I think (1) is the right approach, and I see that you have already opened an issue with Lukas.

Meanwhile, option (3) is to #ifdef out the problematic code as part of the Crux build. We could do that temporarily if Lukas is slow in responding.

**Comment by kfattila** [comment]
Woudn't option (3) be the same as option (2)?

**Comment by wsnoble** [comment]
I just meant to ifdef out the assert part.

**Comment by kfattila** [comment]
Charles has already committed a patch to comment out the assert part. I close this issue.
